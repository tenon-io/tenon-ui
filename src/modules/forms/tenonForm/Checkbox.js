import React, { forwardRef } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import FeedbackBlock from './FeedbackBlock';

/**
 * @component
 * A default view component to use with the tenon-ui
 * smart textinput controller.
 *
 * @prop {string} contentHintText - Text to show as content
 *                hint of the input.
 * @prop {string} errorText - Text to show as error message
 *                of the input.
 * @prop required {function} getLabelProps - Prop getter function
 *              from the smart textinput controller for the <label>.
 * @prop required {function} getCheckboxProps - Prop getter function
 *              from the smart textinput controller for the
 *              <input type="checkbox">.
 * @prop {function} getErrorProps - Prop getter function
 *              from the smart textinput controller for the
 *              error message container.
 * @prop {function} getContentHintProps - Prop getter
 *              function from the smart textinput controller
 *              for the content hint container.
 * @prop required {string} labelText
 * @prop {object} labelProps - Control prop collection
 *              for the <label>
 * @prop {boolean} showError - A flag to indicate when the
 *              error container should be displayed.
 * @prop {string} className - An optional class string to transfer to the
 * className prop of the input element.
 * @prop {string} requiredText - And optional string value to display
 *              after the label indicating a required situation.
 *@prop {boolean} required - An optional boolean value indicating whether
 *              the current view component is attached to a required
 *              data model.
 *
 * NOTE: All props given to this input that does not
 * satisfy one of the above will be passed into the the
 * prop getter function of the input to enrich the input or
 * override already configured default props. Providing
 * extraneous properties that should not be rendered on an
 * <input> in the DOM can create run time errors.
 */
const Checkbox = forwardRef(
    (
        {
            contentHintText,
            errorText,
            getLabelProps,
            getCheckboxProps,
            getErrorProps,
            getContentHintProps,
            labelText,
            labelProps,
            showError,
            className,
            required,
            requiredText,
            ...rest
        },
        ref
    ) => (
        <div className="form-group">
            <div className="checkbox-wrapper">
                <input
                    ref={ref}
                    className={
                        classNames(className, { 'has-error': showError }) ||
                        null
                    }
                    {...getCheckboxProps(rest)}
                />
                <label {...getLabelProps(labelProps)}>
                    {labelText}
                    {required ? (
                        <span aria-hidden="true" className="required">
                            &nbsp;
                            {requiredText || '*'}
                        </span>
                    ) : null}
                </label>
            </div>
            <FeedbackBlock
                getContentHintProps={getContentHintProps}
                getErrorProps={getErrorProps}
                errorText={errorText}
                contentHintText={contentHintText}
                showError={showError}
            />
        </div>
    )
);

Checkbox.displayName = 'Checkbox';

Checkbox.propTypes = {
    contentHintText: PropTypes.string,
    errorText: PropTypes.string,
    getLabelProps: PropTypes.func.isRequired,
    getCheckboxProps: PropTypes.func.isRequired,
    getErrorProps: PropTypes.func,
    getContentHintProps: PropTypes.func,
    labelText: PropTypes.string.isRequired,
    labelProps: PropTypes.object,
    showError: PropTypes.bool,
    className: PropTypes.string,
    required: PropTypes.bool,
    requiredText: PropTypes.string
};

export default Checkbox;
