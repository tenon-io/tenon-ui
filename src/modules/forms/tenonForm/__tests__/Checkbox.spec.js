import React, { createRef } from 'react';
import { cleanup, render } from 'react-testing-library';
import 'jest-dom/extend-expect';
import Checkbox from '../Checkbox';

describe('Checkbox', () => {
    afterEach(cleanup);

    it('should render a checkbox and label only', () => {
        const { container } = render(
            <Checkbox
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getCheckboxProps={() => ({ id: 'foo', type: 'checkbox' })}
                labelText="Test checkbox"
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render an error text when invalid', () => {
        const { container } = render(
            <Checkbox
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getCheckboxProps={() => ({
                    id: 'foo',
                    'aria-describedby': 'bar',
                    type: 'checkbox'
                })}
                getErrorProps={() => ({ id: 'bar' })}
                labelText="Test checkbox"
                errorText="Test error"
                showError={true}
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render a content hint if given', () => {
        const { container } = render(
            <Checkbox
                contentHintText="Content hint test"
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getCheckboxProps={() => ({
                    id: 'foo',
                    'aria-describedby': 'bar',
                    type: 'checkbox'
                })}
                getContentHintProps={() => ({ id: 'bar' })}
                labelText="Test checkbox"
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should allow for easy overriding of default props', () => {
        const { getByLabelText, getByText } = render(
            <Checkbox
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getCheckboxProps={() => ({ id: 'foo', type: 'checkbox' })}
                labelText="Test checkbox"
                labelProps={{ htmlFor: 'foo' }}
            />
        );

        expect(getByLabelText('Test checkbox')).toHaveAttribute('id', 'foo');
        expect(getByText('Test checkbox')).toHaveAttribute('for', 'foo');
    });

    it('should allow easy focus of the internal input', () => {
        const inPutRef = createRef();

        const { getByLabelText } = render(
            <Checkbox
                ref={inPutRef}
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getCheckboxProps={() => ({ id: 'foo', type: 'checkbox' })}
                labelText="Test checkbox"
            />
        );

        expect(getByLabelText('Test checkbox')).toEqual(inPutRef.current);
    });

    it('should render a required text with a fallback', () => {
        const { container, rerender } = render(
            <Checkbox
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getCheckboxProps={() => ({ id: 'foo', type: 'checkbox' })}
                labelText="Test checkbox"
                required={true}
            />
        );

        expect(container.querySelector('span.required').innerHTML).toBe(
            '&nbsp;*'
        );

        rerender(
            <Checkbox
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getCheckboxProps={() => ({ id: 'foo', type: 'checkbox' })}
                labelText="Test checkbox"
                required={true}
                requiredText="required"
            />
        );

        expect(container.querySelector('span.required').innerHTML).toBe(
            '&nbsp;required'
        );
    });
});
