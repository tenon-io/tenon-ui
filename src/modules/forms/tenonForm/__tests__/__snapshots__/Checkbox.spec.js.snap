// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render a checkbox and label only 1`] = `
<div
  class="form-group"
>
  <div
    class="checkbox-wrapper"
  >
    <input
      id="foo"
      type="checkbox"
    />
    <label
      for="foo"
    >
      Test checkbox
    </label>
  </div>
</div>
`;

exports[`Checkbox should render a content hint if given 1`] = `
<div
  class="form-group"
>
  <div
    class="checkbox-wrapper"
  >
    <input
      aria-describedby="bar"
      id="foo"
      type="checkbox"
    />
    <label
      for="foo"
    >
      Test checkbox
    </label>
  </div>
  <div
    class="info-wrapper"
  >
    <span
      class="icon"
    />
    <span
      id="bar"
    >
      Content hint test
    </span>
  </div>
</div>
`;

exports[`Checkbox should render an error text when invalid 1`] = `
<div
  class="form-group"
>
  <div
    class="checkbox-wrapper"
  >
    <input
      aria-describedby="bar"
      class="has-error"
      id="foo"
      type="checkbox"
    />
    <label
      for="foo"
    >
      Test checkbox
    </label>
  </div>
  <div
    class="error-wrapper"
  >
    <span
      class="icon"
    />
    <span
      id="bar"
    >
      Test error
    </span>
  </div>
</div>
`;
